lib/JSON/Validator.pm

```diff
@@ -196,141 +164,133 @@ sub _load_schema_from_text {
-  return $doc;
-}
+  if (ref $schema eq 'HASH') {
+    $id = $schema->{id} // '';
+    return $resolved->schema if $resolved = $self->{refs}{$id};
+  }
+  elsif ($resolved = $self->{refs}{$schema // ''}) {
+    return $resolved->schema;
+  }
+  else {
+    ($schema, $id) = $self->_load_schema($schema);
+    $id = $schema->{id} if $schema->{id};
+  }
 
-# This method is used to resolve all the $ref's that we find inside JSON Schema
-# specification.
-#
-# $namespace is typically the "id" part at the top level of the schema
-# specification. Meaning some kind of URL. In the case of Swagger: Just make
-# something up instead using _default_id()
-sub _resolve_schema {
-  my ($self, $schema, $namespace) = @_;
-  my ($resolved, @topics, @refs);
-
-  # The if is true if we have already resolved this $namespace. This happens if
-  # the same specification has "$ref" with the same value multiple times.
-  # There's no need to resolve the same schema twice.
-  return $resolved if $resolved = $self->_store(resolved => $namespace);
-
-  @topics = ($self->_store(resolved => $namespace, $schema))->data;
-
-  # This while loop will traverse the whole specification and and track down
-  # each and every "$ref" it finds. @topics is a list of all the data
-  # structures that it finds. Note that is start out with just having one item:
-  # A hash-ref to the complete specification, but as it goes along, the while()
-  # loop will track down more objects and arrays inside the specification and
-  # then loop over those as well, util it has visited the whole document.
+  $self->_register_id($id, $schema);
+
+  my @topics = ([$schema, Mojo::URL->new($id)]);
   while (@topics) {
-    my $topic = shift @topics;
-    next if ref $topic and $self->_seen(topic => $topic);    # Avoid recursion
-    if (UNIVERSAL::isa($topic, 'HASH')) {
-      for my $k (sort keys %$topic) {
-        my $v = $topic->{$k};
-        push @topics, $topic->{$k} if ref $topic->{$k};
-        unshift @refs, $topic if $k eq '$ref' and !ref $v;
-      }
+    my ($topic, $base) = @{shift @topics};
+
+    if (UNIVERSAL::isa($topic, 'ARRAY')) {
+      push @topics, map { [$_, $base] } @$topic;
     }
-    elsif (UNIVERSAL::isa($topic, 'ARRAY')) {
-      unshift @topics, @$topic;
+    elsif (UNIVERSAL::isa($topic, 'HASH')) {
+      if ($topic->{'$ref'} and !ref $topic->{'$ref'}) {
+        push @refs, [$topic, $base];
+      }
+      if ($topic->{id} and !ref $topic->{id}) {
+        my $fqn = Mojo::URL->new($topic->{id});
+        $fqn = $fqn->to_abs($base) unless $fqn->is_abs;
+        $self->_register_id($fqn->to_string, $topic);
+      }
+
+      push @topics, map { [$topic->{$_}, $base] } grep { $_ ne '$ref' } keys %$topic;
     }
   }
 
-  # Resolve all the "$ref" we found. This will call _resolve_schema() again.
-  $self->resolver->($self, $namespace, \@refs);
+  for (@refs) {
+    my ($topic, $base) = @$_;
+    $topic->{'$ref'} = $self->_resolve_ref($topic->{'$ref'}, $base);
```

Comment:
I would much rather replace the `$topic` variable in place with a JSON::Validator::Ref object. Do you know how to do that @jberger or if it's a good idea? If that was done, then I would change the while loop in `_validate()` to:

```
while (UNIVERSAL::isa($schema, 'JSON::Validator::Ref')) {
  Carp::confess("Seems like you have a circular \$ref: $schema") if RECURSION_LIMIT < ++$guard;
  $schema = $schema->schema;
}
```